pkg/reconcilermanager/controllers/reconciler_base.go

```diff
@@ -691,3 +723,85 @@ func (r *reconcilerBase) setupOrTeardown(ctx context.Context, syncObj client.Obj
 
 	return nil
 }
+
+// upsertVerticalPodAutoscaler creates or updates the VPA for the reconciler
+// Deployment if the VPA API is enabled and the strategy is either auto or
+// recommend. If the strategy is disabled and the VPA API is enabled, then the
+// VPA will be deleted, if it exists.
+// Returns true if the VPA was created, updated, or already up-to-date.
+func (r *reconcilerBase) upsertVerticalPodAutoscaler(ctx context.Context, strategy metadata.ReconcilerAutoscalingStrategy, reconcilerRef types.NamespacedName, labelMap map[string]string) (client.ObjectKey, bool, error) {
+	vpaRef := reconcilerRef
+	vpaEnabled, err := r.isVPAEnabled()
+	if err != nil {
+		return vpaRef, false, err
+	}
+	switch strategy {
+	case metadata.ReconcilerAutoscalingStrategyDisabled:
+		// delete if VPA is installed
+		if !vpaEnabled {
+			r.logger(ctx).Info("Managed object delete skipped - not enabled",
+				logFieldObjectRef, vpaRef.String(),
+				logFieldObjectKind, "VerticalPodAutoscaler")
+			return vpaRef, false, nil
+		}
+		return vpaRef, false, r.deleteVerticalPodAutoscaler(ctx, reconcilerRef)
+	case metadata.ReconcilerAutoscalingStrategyAuto, metadata.ReconcilerAutoscalingStrategyRecommend:
+		// upsert if VPA is installed
+		if !vpaEnabled {
+			r.logger(ctx).Info("Managed object upsert skipped - not enabled",
+				logFieldObjectRef, vpaRef.String(),
+				logFieldObjectKind, "VerticalPodAutoscaler")
+			return vpaRef, false, nil
+		} // else continue to upsert
+	default:
+		// shouldn't happen - invalid strategy should be caught by validation
+		return vpaRef, false, errors.Errorf("invalid reconciler autoscaling strategy: %v", strategy)
+	}
+	autoscale := (strategy == metadata.ReconcilerAutoscalingStrategyAuto)
+	vpa := &autoscalingv1.VerticalPodAutoscaler{}
+	vpa.Name = vpaRef.Name
+	vpa.Namespace = vpaRef.Namespace
+	op, err := CreateOrUpdate(ctx, r.client, vpa, func() error {
+		r.addLabels(vpa, labelMap)
+		vpa.Spec.TargetRef = &autoscaling.CrossVersionObjectReference{
+			// TODO: APIVersion is optional but is it useful?
+			Kind: kinds.Deployment().Kind,
+			Name: reconcilerRef.Name,
+		}
+		var updateMode autoscalingv1.UpdateMode
+		if autoscale {
```

Comment:
minor nit: I feel like using the boolean here doesn't help much with readability. I feel it would be clearer just comparing directly to the declared consts (e.g. with a `switch`)

Reply:
The Boolean is different than just strategy == Auto here. The Boolean means Auto + APIService exists.

Reply:
oh, this is a different place than I thought you were commenting on. Fixed it here.